File: backend/app/schemas/staff_management.py
"""Enterprise staff management API schemas."""

from __future__ import annotations

from datetime import datetime
from uuid import UUID

from pydantic import BaseModel, ConfigDict, EmailStr, Field

from app.models.enums import PartnerStaffRole


class WorkSchedule(BaseModel):
    model_config = ConfigDict(extra="forbid")

    days: list[str] = Field(default_factory=lambda: ["mon", "tue", "wed", "thu", "fri", "sat"])
    start_time: str = Field(default="09:00", pattern=r"^\d{2}:\d{2}$")
    end_time: str = Field(default="18:00", pattern=r"^\d{2}:\d{2}$")
    timezone: str = Field(default="Asia/Kolkata", max_length=64)


class StaffDashboardResponse(BaseModel):
    laundry_id: UUID
    laundry_name: str
    total_staff: int
    active_staff: int
    online_staff: int
    inactive_staff: int
    suspended_staff: int = 0
    pending_tasks: int
    pending_pickups: int
    pending_deliveries: int
    pending_processing: int


class StaffMemberResponse(BaseModel):
    id: UUID
    laundry_id: UUID
    laundry_name: str
    user_id: UUID | None = None
    name: str
    email: str | None = None
    phone: str | None = None
    role: str
    role_label: str
    is_active: bool
    is_suspended: bool = False
    suspended_reason: str | None = None
    work_schedule: WorkSchedule | None = None
    last_login_at: datetime | None = None
    last_active_at: datetime | None = None
    created_at: datetime
    temporary_password: str | None = None


class StaffCreateRequest(BaseModel):
    model_config = ConfigDict(extra="forbid")

    name: str = Field(min_length=1, max_length=200)
    email: EmailStr
    phone: str | None = Field(default=None, max_length=20)
    role: PartnerStaffRole
    laundry_id: UUID | None = None
    password: str | None = Field(default=None, min_length=8, max_length=128)
    work_schedule: WorkSchedule | None = None


class StaffUpdateRequest(BaseModel):
    model_config = ConfigDict(extra="forbid")

    name: str | None = Field(default=None, min_length=1, max_length=200)
    phone: str | None = Field(default=None, max_length=20)
    role: PartnerStaffRole | None = None
    laundry_id: UUID | None = None
    work_schedule: WorkSchedule | None = None


class StaffSuspendRequest(BaseModel):
    model_config = ConfigDict(extra="forbid")

    reason: str = Field(min_length=3, max_length=500)


class StaffResetPasswordResponse(BaseModel):
    staff_id: UUID
    temporary_password: str


class StaffActivityRow(BaseModel):
    id: UUID
    staff_id: UUID | None = None
    staff_name: str
    action: str
    resource_type: str | None = None
    resource_id: str | None = None
    description: str | None = None
    metadata: dict = Field(default_factory=dict)
    created_at: datetime

File: frontend/services/staff-management.ts
import { api, type ApiEnvelope } from '@/lib/api';



export type StaffRole =

  | 'owner'

  | 'manager'

  | 'pickup_agent'

  | 'delivery_agent'

  | 'operator'

  | 'support_staff';



export interface WorkSchedule {

  days: string[];

  start_time: string;

  end_time: string;

  timezone: string;

}



export const DEFAULT_WORK_SCHEDULE: WorkSchedule = {

  days: ['mon', 'tue', 'wed', 'thu', 'fri', 'sat'],

  start_time: '09:00',

  end_time: '18:00',

  timezone: 'Asia/Kolkata',

};



export interface StaffDashboard {

  laundry_id: string;

  laundry_name: string;

  total_staff: number;

  active_staff: number;

  online_staff: number;

  inactive_staff: number;

  suspended_staff: number;

  pending_tasks: number;

  pending_pickups: number;

  pending_deliveries: number;

  pending_processing: number;

}



export interface StaffMember {

  id: string;

  laundry_id: string;

  laundry_name: string;

  user_id: string | null;

  name: string;

  email: string | null;

  phone: string | null;

  role: StaffRole | string;

  role_label: string;

  is_active: boolean;

  is_suspended: boolean;

  suspended_reason: string | null;

  work_schedule: WorkSchedule | null;

  last_login_at: string | null;

  last_active_at: string | null;

  created_at: string;

  temporary_password?: string | null;

}



export interface StaffActivityRow {

  id: string;

  staff_id: string | null;

  staff_name: string;

  action: string;

  resource_type: string | null;

  resource_id: string | null;

  description: string | null;

  metadata: Record<string, unknown>;

  created_at: string;

}



export interface CreateStaffInput {

  name: string;

  email: string;

  phone?: string;

  role: StaffRole;

  password?: string;

  work_schedule?: WorkSchedule;

}



export interface UpdateStaffInput {

  name?: string;

  phone?: string;

  role?: StaffRole;

  work_schedule?: WorkSchedule;

}



export const STAFF_ROLE_LABELS: Record<string, string> = {

  owner: 'Owner',

  manager: 'Manager',

  pickup_agent: 'Pickup Agent',

  delivery_agent: 'Delivery Agent',

  operator: 'Laundry Operator',

  support_staff: 'Support Staff',

};



export const STAFF_ROLES: StaffRole[] = [

  'manager',

  'pickup_agent',

  'delivery_agent',

  'operator',

  'support_staff',

];



export const WEEKDAYS = [

  { id: 'mon', label: 'Mon' },

  { id: 'tue', label: 'Tue' },

  { id: 'wed', label: 'Wed' },

  { id: 'thu', label: 'Thu' },

  { id: 'fri', label: 'Fri' },

  { id: 'sat', label: 'Sat' },

  { id: 'sun', label: 'Sun' },

];



export async function getStaffDashboard(): Promise<StaffDashboard> {

  const { data } = await api.get<ApiEnvelope<StaffDashboard>>('/partner/staff-management/dashboard');

  return data.data;

}



export async function listStaffMembers(): Promise<StaffMember[]> {

  const { data } = await api.get<ApiEnvelope<StaffMember[]>>('/partner/staff-management');

  return data.data;

}



export async function createStaffMember(body: CreateStaffInput): Promise<StaffMember> {

  const { data } = await api.post<ApiEnvelope<StaffMember>>('/partner/staff-management', body);

  return data.data;

}



export async function updateStaffMember(id: string, body: UpdateStaffInput): Promise<StaffMember> {

  const { data } = await api.patch<ApiEnvelope<StaffMember>>(`/partner/staff-management/${id}`, body);

  return data.data;

}



export async function deactivateStaffMember(id: string): Promise<StaffMember> {

  const { data } = await api.post<ApiEnvelope<StaffMember>>(`/partner/staff-management/${id}/deactivate`);

  return data.data;

}



export async function activateStaffMember(id: string): Promise<StaffMember> {

  const { data } = await api.post<ApiEnvelope<StaffMember>>(`/partner/staff-management/${id}/activate`);

  return data.data;

}



export async function suspendStaffMember(id: string, reason: string): Promise<StaffMember> {

  const { data } = await api.post<ApiEnvelope<StaffMember>>(`/partner/staff-management/${id}/suspend`, { reason });

  return data.data;

}



export async function unsuspendStaffMember(id: string): Promise<StaffMember> {

  const { data } = await api.post<ApiEnvelope<StaffMember>>(`/partner/staff-management/${id}/unsuspend`);

  return data.data;

}



export async function resetStaffPassword(id: string): Promise<{ staff_id: string; temporary_password: string }> {

  const { data } = await api.post<ApiEnvelope<{ staff_id: string; temporary_password: string }>>(

    `/partner/staff-management/${id}/reset-password`,

  );

  return data.data;

}



export async function getStaffActivity(staffId?: string): Promise<StaffActivityRow[]> {

  const { data } = await api.get<ApiEnvelope<StaffActivityRow[]>>('/partner/staff-management/activity', {

    params: staffId ? { staff_id: staffId } : undefined,

  });

  return data.data;

}



export const ACTIVITY_LABELS: Record<string, string> = {

  login: 'Login',

  logout: 'Logout',

  order_update: 'Order update',

  assignment: 'Assignment',

  status_change: 'Status change',

  staff_created: 'Staff created',

  staff_updated: 'Staff updated',

  staff_deactivated: 'Staff deactivated',

  staff_suspended: 'Suspended',

  staff_unsuspended: 'Unsuspended',

  password_reset: 'Password reset',

};



export function formatWorkSchedule(schedule: WorkSchedule | null | undefined): string {

  if (!schedule) return '—';

  const dayLabels = schedule.days.map((d) => WEEKDAYS.find((w) => w.id === d)?.label ?? d).join(', ');

  return `${dayLabels} · ${schedule.start_time}–${schedule.end_time}`;

}

